src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,296 @@
+                                Right: ThrowExpressionSyntax { Expression: ObjectCreationExpressionSyntax thrownInNullCoalescing }
+                            }
+                        }
+                    }:
+                        var coalescedParameter = TryGetParameter(maybeParameter);
+                        if (coalescedParameter is null || !IsConstructorApplicable(thrownInNullCoalescing, coalescedParameter))
+                        {
+                            return null;
+                        }
+
+                        return coalescedParameter;
+
+                    default:
+                        return null;
+                }
+            }
+
+            IParameterSymbol? TryGetParameter(ExpressionSyntax maybeParameter)
+            {
+                // `(object)x == null` is often used to ensure reference equality is used.
+                // therefore, we specially unwrap casts when the cast is to `object`.
+                if (maybeParameter is CastExpressionSyntax { Type: var type, Expression: var operand })
+                {
+                    if (semanticModel.GetTypeInfo(type).Type?.SpecialType != SpecialType.System_Object)
+                    {
+                        return null;
+                    }
+
+                    maybeParameter = operand;
+                }
+
+                if (semanticModel.GetSymbolInfo(maybeParameter).Symbol is not IParameterSymbol { ContainingSymbol: { } containingSymbol } parameterSymbol || !containingSymbol.Equals(methodSymbol))
+                {
+                    return null;
+                }
+
+                return parameterSymbol;
+            }
+
+            bool IsConstructorApplicable(ObjectCreationExpressionSyntax exceptionCreation, IParameterSymbol parameterSymbol)
+            {
+                if (exceptionCreation.ArgumentList?.Arguments is not { } arguments)
+                {
+                    return false;
+                }
+
+                if (arguments.Count == 0)
+                {
+                    // 'new ArgumentNullException()'
+                    return argumentNullExceptionConstructor.Equals(semanticModel.GetSymbolInfo(exceptionCreation, cancellationToken).Symbol);
+                }
+
+                if (arguments.Count != 1)
+                {
+                    return false;
+                }
+
+                // 'new ArgumentNullException(nameof(param))' (or equivalent)
+                var constantValue = semanticModel.GetConstantValue(arguments[0].Expression, cancellationToken);
+                if (constantValue.Value is not string constantString || !string.Equals(constantString, parameterSymbol.Name, StringComparison.Ordinal))
+                {
+                    return false;
+                }
+
+                return argumentNullExceptionStringConstructor.Equals(semanticModel.GetSymbolInfo(exceptionCreation, cancellationToken).Symbol);
```

Comment:
consider reordering.  it's a personal nit, but it seems more sensible to first see if you're doing `new ArgumntNullException` *then* check the argument.